### ISSUE TITLE
Check that setup.py correctly installs dependencies (and has them all)

### ISSUE THREAD

**Issue by jkitzes** [jkitzes/macroeco#92]

**Comment by jkitzes** [comment]
Make sure can do clean install on Linux and Windows

**Comment by jkitzes** [comment]
@mqwilber , I believe that the current `develop` branch is ready to go in this regard. Can you test on a few clean machines to make sure the dependencies install correctly? I'll test on this end also on Mac and Linux - if you can find a Windows machine without Python on it and try to install, that would be ideal.

Here's the relevant draft installation instructions from the manuscript - to use the pip command I'll have to upload to PyPI, so hold on for a bit until I get that taken care of -

For users wishing to install a new scientific Python environment from scratch, we recommend installing the Miniconda Python 2.7 distribution from Continuum Analytics (http://conda.pydata.org/ miniconda.html). After installing Miniconda, run the command `conda install numpy scipy pandas matplotlib pip`, followed by the command `pip install macroeco`, to complete the installation of _macroeco_. These commands should be run from a terminal window in Mac OS X or Linux or from the command prompt in Windows.

Users wishing to use the _MacroecoDesktop_ graphical interface will also need to install the optional dependency _wxpython_. Miniconda users can install this package with the command `conda install wxpython`.

**Comment by mqwilber** [comment]
@jkitzes, sounds good. Let me know when you have uploaded to PyPi

**Comment by jkitzes** [comment]
Well, that was harder than I thought (all sorts of bugs on Linux), but I think it's done now - so @mqwilber, what would be great is if you could do the following on a clean Mac, Windows, and Linux machine with no existing Python install (or on one where you can clobber it):

1. Install Miniconda
2. Run `conda install numpy scipy pandas matplotlib pip`
3. Run `pip install macroeco`
4. Make sure `macroeco` imports in an interpreter and run a few examples lines, then make sure that you can run both of the commands below from the command line:

`python -c "import macroeco; macroeco.desktop()"`

`python -c "import macroeco; macroeco.desktop('\path\to\parameters.txt)"`

and that these both correctly execute a MacroecoDesktop analysis.

I've tried on all 3 platforms over here and it works, but a second set of eyes would be useful..

**Comment by mqwilber** [comment]
Tried it on two naive windows machines.  Here are the problems I encountered.

On a Windows Vista Machine:

1. `pip install macroeco` gave the following error: `error: invalid command 'bdist_wheel`

**Comment by mqwilber** [comment]
Tried it on two naive windows machines.  Here are the problems I encountered.

**On a Windows Vista Machine**

1. `pip install macroeco` gave the following error: `error: invalid command 'bdist_wheel`
   - **Solution**: For some reason I needed to `pip uninstall wheel` and then `pip install wheel` and then `macroeco` loaded fine

2. `ImportError: no packaged named wx`
    - **Solution**: I needed to manually install wxPython from http://www.wxpython.org/download.php

After those two hiccups, things were working well on the command line

3. I ran `python -c "import macroeco; macroeco.desktop()"` the Desktop GUI opened file.  However, when executing the demo `parameters.txt` I got the following error

`WindowsError: [Error 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Users\\Administrator\\Desktop\\results\\SAD ANBO Row > 2`

Same error when I run `python -c "import macroeco; macroeco.desktop('\path\to\parameters.txt')"`

   - **Solution**: the `>` sign is making Windows mad. Change to `SAD ANBO Row greater than 2` and the analysis works fine

4. Trying to re-save the parameters file and run another analysis on Windows Vista gives the following error

`WindowsError: [Error 32] The process cannot access the file because it is being used by another process: `_log.txt`

   - **Solution**: Did not find one other than closing and reopening the GUI.

**On Linux**: Coming soon...

**Comment by jkitzes** [comment]
Great, thanks for all of the above.

1. The wheel issue is strange, I'll try to replicate it. I haven't been able to so far...

2. Whoops, mistake in the installation line, should be `conda install numpy scipy pandas matplotlib pip` which takes care of the wxpython issue.

3. OK, so we need to eliminate special chars in the section names, I'll put that in the manuscript.

4. That means the previous analysis had hung and not closed I think, so closing and opening the GUI makes sense. If it hadn't crashed the first time on the parameter file then this shouldn't have been a problem.

The wheel issue concerns me since it seems to be intermittent...

**Comment by jkitzes** [comment]
On wheel, on my Mac the `conda install pip` installs `wheel` as a dependency which seemed to work fine, don't know why on Windows it wouldn't have. I just tried in a virtualenv on an existing Windows machine and there was no wheel problem (Windows 7).

**Comment by mqwilber** [comment]
Might be a Vista thing. `wheel` was installed, but for some reason I had to uninstall it and then reinstall it for it to work. Trying to replicate it on another Windows machine.

**Comment by jkitzes** [comment]
I don't know if you're going to be able to replicate the issue on that Vista machine, but if you just uninstall or erase all of Miniconda, you might try the same instructions again (including the wxpython part of the first conda command). If you get the same wheel error, try running

`pip install -i https://testpypi.python.org/pypi macroeco`

instead of

`pip install macroeco`

and see if there's a chance the first one might work where the second one fails. This is all I can come up with for the moment. Of course we could also just say that the uninstall/reinstall of wheel may be necessary, but it would be nice if we didn't have to.

**Comment by mqwilber** [comment]
How is 
> Whoops, mistake in the installation line, should be conda install numpy scipy pandas matplotlib pip which takes care of the wxpython issue.

different than what was given above?

**Comment by jkitzes** [comment]
Ha, sorry - use 'conda install numpy scipy pandas matplotlib pip wxpython` instead of `conda install numpy scipy pandas matplotlib pip` (just add the wxpython argument too).

**Comment by mqwilber** [comment]
Perfect, thanks.  Things are worked perfectly on the Linux test.

**Comment by mqwilber** [comment]
Ok. I cleared Miniconda from the vista machine a reloaded everything. No errors this time around. 

However, I am still getting `WindowsError: [Error 32] The process cannot access the file because it is being used by another process:` on the GUI MacroecoDesktop when rerunning an analysis - even after a successful parameters file run.  Did you get this error on your end @jkitzes?

**Comment by jkitzes** [comment]
Great - to confirm on the Windows wheel issue, it just worked this time, didn't do anything differently? You still used `pip install macroeco` for the second Vista installation, not the longer command with the testpypi in it above?

Yes, I can confirm the WindowsError - apparently when the thread that's running the actual MacroecoDesktop analysis closes on Mac and Linux, it releases the lock on the log file, but not in windows. Adding a line to close it explicitly seems to do the trick. Try uninstalling `macroeco` then running

`pip install -i https://testpypi.python.org/pypi macroeco`

which will install a version 0.99 that should take care of this. If you confirm, I'll push it to main PyPI site. Annoyingly, since I've already pushed up 1.0, I can't actually replace those files, so we'll have to call this 1.0.1.

**Comment by mqwilber** [comment]
Ok, almost there.  I can now run multiple parameter analyses without a WindowsError if they all execute correctly.  However, if the GUI throws an error (a WindowsError or another error) and then I try another run I am still getting `WindowsError` with a lock on the `_log.txt`.

**Comment by jkitzes** [comment]
I'm glad that you're testing much more thoroughly than I am (apparently)! OK think I got it, put a command to shutdown logging in the GUI code as well, so before the button Save and Run comes back on, it closes the log file. Try `pip uninstall macroeco` and then `pip install -i https://testpypi.python.org/pypi macroeco` again - you should get v0.99.1.

**Comment by jkitzes** [comment]
Also, think we need to worry about the `wheel` package issue?

**Comment by mqwilber** [comment]
Just tested it and v0.99.1 seemed to fix the problem.  

I am encountering one difference between `pip install -i https://testpypi.python.org/pypi macroeco` (v0.99.1) and `pip install macroeco` (v1.0).  The former will not automatically install the necessary dependencies (i.e. `configparser`, `mpmath`, `decorator`) so you have to do them by manually (i.e. `pip install configparser`).  Using `pip install macroeco` everything is installed automatically (though you of course get v1.0). This may be function of the testpypi setup, but let's check on this when you push v1.0.1 to PyPi. 

I didn't encounter the `wheel` issue on this most recent test, so I don't think we need to worry about it.  We could just place a note in the documentation describing what todo if you encounter this problem upon installing.

**Comment by jkitzes** [comment]
Right, that makes sense about the install, since the dependencies aren't going to be in testpypi. You can test the whole thing including dependencies with `pip install --extra-index-url https://testpypi.python.org/pypi macroeco` instead, that should do everything correctly.

OK on `wheel`, we'll leave it off for now. Mostly I hope it doesn't suddenly appear for our manuscript reviewers...

**Comment by mqwilber** [comment]
Tested on a Windows 10 machine as well and did not get a `wheel` error.
